text/0000-stable-simd.md

```diff
@@ -0,0 +1,720 @@
+with a future const-powered world of Rust but in the meantime not otherwise
+block stabilization of these intrinsics.
+
+## Portable packed SIMD
+
+So-called "portable" packed SIMD types are currently implemented in both the
+[stdsimd] and [simd][simd-crate] crates. These types look like `u8x16` and
+explicitly specify how many lanes they have (16 in this case) and what type each
+line is (`u8` in this case). These types are intended to unconditionally
+available (like the rest of libstd) and simply optimized much more aggressively
+on platforms that have native support for the various operations.
+
+For example `u8x16::add` may be implemented differently on i586 vs i686, and
+also entirely differently implemented on ARM. The idea with portable packed SIMD
+types is that they represent a broad intersection of fast behavior across a
+broad range of platforms.
+
+It's intended that this RFC neither includes nor rules out the addition of
+portable packed-SIMD types in Rust. It's expected that an upcoming RFC will
+propose the addition of these types in a `std::simd` module. These types will be
+orthogonal to scalable-vector types which are expected to be proposed in
+another, also different, RFC.  What this RFC does do, however, is explicitly
+specify that:
+
+* The portable SIMD types (both packed and scalable) will not be used in vendor
+  intrinsics.
+* The vendor SIMD types will be distinct types from the portable SIMD types.
+
+Or, in other words, it's intended that portable SIMD types are entirely
+decoupled from vendor intrinsics. If they both end up being implemented then
+there will be zero-cost interoperation between them, but neither
+will necessarily depend on the other.
+
+[soundbug]: https://github.com/rust-lang/rust/issues/44367
+
+## Not stabilizing MMX in this RFC
+
+This RFC proposed notably omitting the MMX intrinsics, or those related to
+`__m64` in other words. The MMX type `__m64` and the intrinsics have been
+somewhat problematic in a number of ways. Known cases include:
+
+* [MMX intrinsics aren't always desirable][mmx1]
+* [LLVM codegen errors happen with debuginfo enabled and MMX][mmx2]
+* [LLVM codegen errors with MMX types and i586][mmx3]
+
+[mmx1]: https://github.com/rust-lang/rust/pull/45367#issuecomment-337883136
+[mmx2]: https://github.com/rust-lang-nursery/stdsimd/issues/246
+[mmx3]: https://github.com/rust-lang-nursery/stdsimd/issues/300
+
+Due to these issues having an unclear conclusion as well as a seeming lack of
+desire to stabilize MMX intrinsics, the `__m64` and all related intrinsics
+**will not be stabilized** via this RFC.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+This RFC represents a *significant* addition to the standard library, maybe one
+of the largest we've ever done! As a result alternate implementations of Rust
+will likely have a difficult time catching up to rustc/LLVM with all the SIMD
+intrinsics. Additionaly the semantics of "packed SIMD types should work everywhere"
```

Comment:
It's worth noting that the vast majority of these intrinsics (especially the SIMD ones) can be emulated with scalar code, and with relatively little code per intrinsic (vendors often even give scalar pseudo-code detailing the semantics of each instructions). While this is still far from trivial due to the large number of intrinsics, it's probably better than having to add first-class support for all these intrinsics to the backend. For example, cretonne probably has no support for `_mm_abs_epi8` right now, and writing a scalar emulation may be easier than adding that support &ndash; the effort can be shared between multiple backends/Rust compilers, and it doesn't dealing with the instruction encodings of the `pabsb` instruction (as well as any other machine-level information that cretonne may need to know).

Reply:
(For cretonne in rustc specifically, we could also just punt on these intrinsics and fall back to LLVM, like we'd presumably do for other features like inline asm. But the general point still stands.)

Reply:
I'm having a hard time parsing this into a concrete suggestion. Could you spell it out a little more for me? :)

Reply:
I'm not proposing anything at this time since this drawback is not a practical issue right now. I'm just pointing out: if we ever end up in a situation where we have alternative Rust implementations that are seriously affected by this drawback, there's two options: (1) Adding real support for all the intrinsics to that compiler, with all that entails, or (2) emulating as many intrinsics as possible with ordinary Rust code. The latter seems like it should be somewhat easier, alleviating this drawback to some degree.

Reply:
I have a very vague proposal along the lines of "teach Cretonne the entire unprivileged ISA and then expose that to the compiler, with special mappings only for LLVM", but we'll see where we end up.

Reply:
@rkruppe Oh I see! I wasn't sure. But yeah that definitely make sense. Thanks for spelling it out.